[Pluggable components] Error handling documentation (#3067)
dapr/docs: daprdocs/content/en/developing-applications/develop-components/pluggable-components/develop-pluggable.md

This concrete implementation and auxiliary code are the **core** of your pluggable component. They define how your component behaves when handling gRPC requests from Dapr.

## Return semantic state store errors

In order to meet concurrency requirements, your state store pluggable components need error mapping from gRPC returned errors to semantic errors. Each of the three Dapr state store errors below maps to a gRPC status code. 

| Error                    | gRPC error code                 | 
| ------------------------ | ------------------------------- | 

> | ETag Mismatch            | `GRPCCodeETagMismatch`          | 
> | ETag Invalid             | `GRPCCodeETagInvalid`           | 
> | Bulk Delete Row Mismatch | `GRPCCodeBulkDeleteRowMismatch` | 

Reviewer: these are the variable names you should use gRPC codes instead
```suggestion
| ETag Mismatch            | `codes.FailedPrecondition`          | 
| ETag Invalid             | `codes.InvalidArgument`           | 
| Bulk Delete Row Mismatch | `codes.Internal` | 
```

Learn more about concurrency requirements in the [State Management overview]({{< ref "state-management-overview.md#concurrency" >}}).

The following examples demonstrate how to return an error in your own pluggable component, changing the messages to suit your needs.

{{< tabs ".NET" "Java" "Go" >}}
 <!-- .NET -->
{{% codetab %}}

**Etag Mismatch**

```csharp

```

**Etag Invalid**

```csharp

```

**Bulk Delete Row Mismatch**

```csharp

```

{{% /codetab %}}

 <!-- Java -->
{{% codetab %}}

**ETag Mismatch**

```java

```

**ETag Invalid**

```java

```

**Bulk Delete Row Mismatch**

```java

```

{{% /codetab %}}

 <!-- Go -->
{{% codetab %}}

**ETag Mismatch**

```go

> st := status.New(GRPCCodeETagMismatch, "fake-err-msg")

Reviewer: ```suggestion
st := status.New(codes.FailedPrecondition, "fake-err-msg")
```

desc := "The ETag field must only contain alphanumeric characters"
v := &errdetails.BadRequest_FieldViolation{
	Field:       etagField,
	Description: desc,
}
br := &errdetails.BadRequest{}
br.FieldViolations = append(br.FieldViolations, v)
st, err := st.WithDetails(br)
```

**ETag Invalid**

```go

> st := status.New(GRPCCodeETagInvalid, "fake-err-msg")

Reviewer: ```suggestion
st := status.New(codes.InvalidArgument, "fake-err-msg")
```

desc := "The ETag field must only contain alphanumeric characters"
v := &errdetails.BadRequest_FieldViolation{
	Field:       etagField,
	Description: desc,
}
br := &errdetails.BadRequest{}
br.FieldViolations = append(br.FieldViolations, v)
st, err := st.WithDetails(br)
```

**Bulk Delete Row Mismatch**

```go

> st := status.New(GRPCCodeBulkDeleteRowMismatch, "fake-err-msg")

Reviewer: ```suggestion
st := status.New(codes.Internal, "fake-err-msg")
```

br := &errdetails.ErrorInfo{}
br.Metadata = map[string]string{

> 	affectedRowsMetadataKey: "100",
> 	expectedRowsMetadataKey: "99",

Reviewer: ```suggestion
	"affected": "100",
	"expected": "99",
```

This concrete implementation and auxiliary code are the **core** of your pluggable component. They define how your component behaves when handling gRPC requests from Dapr.

> ## Return semantic state store errors

Reviewer: Although we don't have any other errors than those returned by state store components, I think we should name this section more broadly e.g `Returning semantic errors`.

And then, we can add one more column in the table below specifying the source component, how about that?

Author: Sounds good to me

> In order to meet concurrency requirements, your state store pluggable components need error mapping, from gRPC returned errors to semantic errors. Each of the three Dapr state store errors below maps to a gRPC status code. 

Reviewer: This description is targeting EtagMismatch error only, as we have others I think we can use this as a `Description` column, how about?

Reviewer: I would say that we should use this initial paragraph to describe the goal of returning semantic errors. e.g

```suggestion
Returning semantic errors are also part of the pluggable component protocol. The component must return specific gRPC codes that have semantic meaning for the user application, those errors are used to a variety of situations from concurrency requirements to informational only.
```

| Error                    | gRPC error code                 | 
| ------------------------ | ------------------------------- | 
| ETag Mismatch            | `codes.FailedPrecondition`          | 
| ETag Invalid             | `codes.InvalidArgument`           | 
| Bulk Delete Row Mismatch | `codes.Internal` | 

Learn more about concurrency requirements in the [State Management overview]({{< ref "state-management-overview.md#concurrency" >}}).

The following examples demonstrate how to return an error in your own pluggable component, changing the messages to suit your needs.

{{< tabs ".NET" "Java" "Go" >}}
 <!-- .NET -->
{{% codetab %}}

> **Important:** In order to use .NET for error mapping, first install the [`Google.Api.CommonProtos` NuGet package](https://www.nuget.org/packages/Google.Api.CommonProtos/).

**Etag Mismatch**

```csharp
var badRequest = new BadRequest();

> var des = "The ETag field must only contain alphanumeric characters";

Reviewer 2: Is this error message correct? It is the same as the ETag Invalid.

```suggestion
var des = "The ETag field provided does not match the one in the store";
```

Reviewer: good catch

badRequest.FieldViolations.Add(
   new Google.Rpc.BadRequest.Types.FieldViolation
   {
      Field = "etag",
      Description = des
   });

var baseStatusCode = Grpc.Core.StatusCode.InvalidArgument;
var status = new Google.Rpc.Status
{
   Code = (int)baseStatusCode
};

status.Details.Add(Google.Protobuf.WellKnownTypes.Any.Pack(badRequest));

var metadata = new Metadata();
metadata.Add("grpc-status-details-bin", status.ToByteArray());
throw new RpcException(new Grpc.Core.Status(baseStatusCode, "fake-err-msg"), metadata);
```

**Bulk Delete Row Mismatch**

```csharp
var errorInfo = new Google.Rpc.ErrorInfo();

errorInfo.Metadata.Add("expected", "100");
errorInfo.Metadata.Add("affected", "99");

var baseStatusCode = Grpc.Core.StatusCode.Internal;
var status = new Google.Rpc.Status{
    Code = (int)baseStatusCode
};

status.Details.Add(Google.Protobuf.WellKnownTypes.Any.Pack(errorInfo));

var metadata = new Metadata();
metadata.Add("grpc-status-details-bin", status.ToByteArray());
throw new RpcException(new Grpc.Core.Status(baseStatusCode, "fake-err-msg"), metadata);
```

{{% /codetab %}}

 <!-- Java -->
{{% codetab %}}

> 

Reviewer 2: ```suggestion

Just like [Dapr Java SDK](https://github.com/tmacam/dapr-java-sdk/), Java Pluggable Components SDK uses [Project Reactor](https://projectreactor.io/), which provides an asynchronous API for Java.

Errors can be returned directly calling the `.error()` method in the `Mono` of `Flux` your method returns and providing the appropriate exception as parameter. You can also raise an exception, as long as it is captured and fed back to your resulting `Mono` or `Flux`.

```

**ETag Mismatch**

```java

> 

Reviewer 2: ```suggestion
    final Status status = Status.newBuilder()
        .setCode(io.grpc.Status.Code.FAILED_PRECONDITION.value())
        .setMessage("fake-err-msg-for-etag-mismatch")
        .addDetails(Any.pack(BadRequest.FieldViolation.newBuilder()
            .setField("etag")
            .setDescription("The ETag field provided does not match the one in the store")
            .build()))
        .build();
    return Mono.error(StatusProto.toStatusException(status));
```

```

**ETag Invalid**

```java

> 

Reviewer 2: ```suggestion
    final Status status = Status.newBuilder()
        .setCode(io.grpc.Status.Code.INVALID_ARGUMENT.value())
        .setMessage("fake-err-msg-for-invalid-etag")
        .addDetails(Any.pack(BadRequest.FieldViolation.newBuilder()
            .setField("etag")
            .setDescription("The ETag field must only contain alphanumeric characters")
            .build()))
        .build();
    return Mono.error(StatusProto.toStatusException(status));
```

```

**Bulk Delete Row Mismatch**

```java

> 

Reviewer 2: ```suggestion
    final Status status = Status.newBuilder()
        .setCode(io.grpc.Status.Code.INTERNAL.value())
        .setMessage("fake-err-msg-for-bulk-delete-row-mismatch")
        .addDetails(Any.pack(ErrorInfo.newBuilder()
            .putAllMetadata(Map.ofEntries(
                Map.entry("affected", "100"),
                Map.entry("expected", "99")
            ))
            .build()))
        .build();
    return Mono.error(StatusProto.toStatusException(status));
```

Reviewer 2: Although I think that the values for affected and expected are swapped, I prefer to follow the example of the other code samples just for the sake of consistency. We might want to verify if that is correct and fix for the 3 languages.

```

{{% /codetab %}}

 <!-- Go -->
{{% codetab %}}

**ETag Mismatch**

```go
st := status.New(codes.FailedPrecondition, "fake-err-msg")

> desc := "The ETag field must only contain alphanumeric characters"

Reviewer 2: I think we are using the wrong message to describe etag mismatch.

```suggestion
desc := "The ETag field provided does not match the one in the store"
```

v := &errdetails.BadRequest_FieldViolation{
	Field:       etagField,
	Description: desc,
}
br := &errdetails.BadRequest{}
br.FieldViolations = append(br.FieldViolations, v)
st, err := st.WithDetails(br)
```

**ETag Invalid**

```go
st := status.New(codes.InvalidArgument, "fake-err-msg")
desc := "The ETag field must only contain alphanumeric characters"
v := &errdetails.BadRequest_FieldViolation{
	Field:       etagField,
	Description: desc,
}
br := &errdetails.BadRequest{}
br.FieldViolations = append(br.FieldViolations, v)
st, err := st.WithDetails(br)
```

**Bulk Delete Row Mismatch**

```go
st := status.New(codes.Internal, "fake-err-msg")
br := &errdetails.ErrorInfo{}
br.Metadata = map[string]string{
	affected: "100",
	expected: "99",

> }

Reviewer 2: @mcandeia Isn't this backwards? Like, we **expected** this delete to impact **100** but but 1 failed and thus only 99 were affected?

If so, other samples will need to be updated too

Reviewer: True, good catch
